### DISCUSSION TITLE
I tried to buld a yolo 11 model for object detection but not getting intended result. Im not sure where I went wrong. Im very new to this.

### DISCUSSION THREAD

**Discussion by sekhrana** [ultralytics/ultralytics#20917] [Q&A]
I tried to train images to find all the pipework in the images. But the the detection has been very bad. I ran it into google collab with T4 and the prediction has been very bad. I have attached the images of the actual annotated image and the predicted image.
this is what I used to train my model 
model = YOLO("yolo11s.yaml")
results = model.train(data=os.path.join(ROOT_DIR, 'pipework_desc.yaml'), epochs=100).
I have 660 images divided into validation and training. 

Are there any obvious issues with my training setup or configuration?
Is my dataset size (660 images) sufficient for this task, or should I collect more data?
Could the problem be related to annotation quality?
Any tips for improving detection performance for complex structures like pipework? 
Thanks for your answers. I new student and learning.
![val_batch0_labels](https://github.com/user-attachments/assets/d338e173-5b9b-4dc3-950f-c3f03824dd34)
![val_batch0_pred](https://github.com/user-attachments/assets/a72ede82-8c20-4a77-99b6-216f91b38305)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sekhrana, thank you for sharing your experience and for your interest in Ultralytics 🚀! This is an automated response to help guide you while an Ultralytics engineer reviews your discussion and provides more targeted assistance soon.

For new users, we recommend visiting the [Docs](https://docs.ultralytics.com/) where you can find [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage guides, along with answers to many common questions.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) with code, data samples, and error logs to help us debug effectively.

For training or performance ❓ questions, please provide as much detail as possible, such as:
- Dataset image and annotation examples (which you’ve provided—thank you!)
- Training logs
- Confirmation that you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

You can join the Ultralytics community for support and peer discussions:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to share tips and connect

## Upgrade

First, ensure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for reaching out and sharing your details! An Ultralytics engineer will follow up with you soon to provide further help.

**Reply by glenn-jocher** [reply]
Hello @sekhrana! Looking at your training setup, I can see a few key areas that could significantly improve your results:

**Use Pretrained Weights**: You're currently building from scratch with `yolo11s.yaml`. Instead, try `model = YOLO("yolo11s.pt")` to start with pretrained COCO weights - this will give you a much better foundation for learning pipework features.

**Dataset Size**: 660 images is on the smaller side for complex industrial detection. For pipework specifically, consider collecting more diverse images with different lighting conditions, angles, and pipe types/materials. Aim for 1000+ images if possible.

**Training Configuration**: Add these parameters to improve training:
```python
results = model.train(
    data='pipework_desc.yaml', 
    epochs=200,  # More epochs with pretrained weights
    imgsz=640,
    patience=50,  # Early stopping
    save_period=10  # Save checkpoints
)
```

**Annotation Quality**: Ensure your bounding boxes tightly fit the pipe sections and consider breaking long continuous pipes into smaller, distinct segments rather than one large box.

Could you share your training logs or the contents of your `pipework_desc.yaml` file? This would help identify any data configuration issues. Also check our [training tips guide](https://docs.ultralytics.com/guides/model-training-tips/) for more optimization strategies.

**Reply by sekhrana** [reply]
Thanks for the reply , I attaching all the log files generated after the training the model.
![confusion_matrix](https://github.com/user-attachments/assets/c5d7ba15-64d0-42ee-bb8f-5c983313d640)
![confusion_matrix_normalized](https://github.com/user-attachments/assets/a8f19232-99f0-4a43-9dff-50548d9243f5)
![F1_curve](https://github.com/user-attachments/assets/d1e9e524-c6ae-4b60-aa3c-24bb8fd7cfa2)
![labels](https://github.com/user-attachments/assets/c8219aea-9466-45ff-a203-ca316bd71575)
![labels_correlogram](https://github.com/user-attachments/assets/40de916e-9f75-41ca-a717-1466cbf19c4b)
![P_curve](https://github.com/user-attachments/assets/0f8f0551-54fc-4293-8c8a-2a4e26fd91db)
![PR_curve](https://github.com/user-attachments/assets/dbee7133-ff6b-4817-9796-f578e5d19e30)
![R_curve](https://github.com/user-attachments/assets/c183f677-dc42-4ee4-a756-5afacd5c269d)
[results.csv](https://github.com/user-attachments/files/20594669/results.csv)
![results](https://github.com/user-attachments/assets/3e89794f-651c-4341-b1ab-38c30c59b324)

**Reply by sekhrana** [reply]
train: /content/drive/My Drive/pipework/train
val: /content/drive/My Drive/pipework/val

nc: 1
names: ['Pipework']  

Contents of the pipework_desc.yaml file.
![val_batch2_labels](https://github.com/user-attachments/assets/b96e46ce-4329-4387-bdc8-9b8390dc0aa6)
![val_batch2_pred](https://github.com/user-attachments/assets/391ecb18-af6f-446b-9429-974655f22c7a)

**Reply by sekhrana** [reply]
Since the pipework is not straight , I have to make annotations smaller so as to not add other equipments as pipework in the bounding box, is it okay to do so. Is this correct way of doing it, but , it includes lot of other equipments.
![test1](https://github.com/user-attachments/assets/d8a0d0be-8677-41f1-a990-477822ef778e)

**Reply by Y-T-G** [reply]
Your annotations are incorrect. You should be using one annotation for a whole single pipe. It doesn't matter if the box includes the background.

**Comment by seopseopi** [comment]
The pretrained-checkpoint change was already covered above. The unresolved question is **what one `Pipework` box means**. In your [test image](https://github.com/user-attachments/assets/d8a0d0be-8677-41f1-a990-477822ef778e), several red rectangles overlap around the bend and vertical run; from that image alone, I cannot tell whether they represent separate physical pipes or short pieces of the same pipe.

- If the target is **one physical pipe**, draw one box around that pipe's visible extent, even though an axis-aligned box includes background and other equipment. A bend by itself does not make a new instance. Split only where you have decided a different pipe begins or where occlusion makes continuity impossible to determine.
- If the target is **a local pipe section**, define repeatable boundaries (for example, between joints or elbows) and label sections that way in *every* image. Then evaluate the model as a section detector, not as a detector that should return one box for the whole pipe.

I would first review 20–30 **original, unaugmented** train and validation images, trace each visible pipe, and check for duplicate boxes on the same chosen unit, missing units, and cases where one image uses whole pipes while another uses fragments. Correct both splits to the same rule before retraining. The crowded validation mosaics make those checks hard to do by eye. If the actual requirement is to isolate the pipe's curved pixels without including equipment, a [segmentation mask](https://docs.ultralytics.com/tasks/segment/) is a better target than a rectangular detection box. Ultralytics' [annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/) also emphasizes defining and applying one labeling rule consistently.

**Reply by pderrenger** [reply]
@sekhrana, should one `Pipework` box represent an entire physical pipe or a section of it? That’s the key question before changing the training setup or collecting more images. Could you share one original image with its annotations and indicate whether the overlapping boxes mark different pipes or sections of the same pipe?
